packages/alchemy/src/middleware/gas-fees.ts

```diff
@@ -0,0 +1,56 @@
+import type { AlchemyProvider } from "../provider";
+
+export enum GasFeeStrategy {
+  DEFAULT = "DEFAULT",
+  FIXED = "FIXED",
+  BASE_FEE_PERCENTAGE = "BASE_FEE_PERCENTAGE",
+  PRIORITY_FEE_PERCENTAGE = "PRIORITY_FEE_PERCENTAGE",
+}
+
+export interface GasFeeMode {
+  strategy: GasFeeStrategy;
+  value: bigint;
+}
+
+export const withAlchemyGasFeeEstimator = (
+  provider: AlchemyProvider,
+  feeMode: GasFeeMode
+): AlchemyProvider => {
+  if (feeMode.strategy === GasFeeStrategy.DEFAULT) {
+    return provider;
+  }
+
+  provider.withFeeDataGetter(async () => {
+    const block = await provider.rpcClient.getBlock({ blockTag: "latest" });
+    const baseFeePerGas = block.baseFeePerGas;
+    if (baseFeePerGas == null) {
+      throw new Error("baseFeePerGas is null");
+    }
+    const maxPriorityFeePerGas = BigInt(
+      await provider.rpcClient.getMaxPriorityFeePerGas()
+    );
+    // add 25% overhead to ensure mine
+    // because we're using bigint, 5n/4n = 1, so we need to scale by 100
+    const baseFeeScaled = (baseFeePerGas * 125n) / 100n;
+
+    const prioFee = ((): bigint => {
+      switch (feeMode.strategy) {
+        case GasFeeStrategy.FIXED:
+          return maxPriorityFeePerGas + feeMode.value;
+        case GasFeeStrategy.BASE_FEE_PERCENTAGE:
+          return (baseFeeScaled * feeMode.value) / 100n;
+        case GasFeeStrategy.PRIORITY_FEE_PERCENTAGE:
+          // add 10% to required priority fee to ensure mine
```

Comment:
Not for this PR, but a general comment. These arbitrary % increases were mostly random guesses by me. We should strive to provide more flexibility for users.

The solution here that @dphilipson and I discussed is some sort of `rundler_feeHistory` endpoint that allows users to estimate their base/priority fees based on prior value and how quickly they want their UO mined. Similar to those fast/normal/slow options that wallets sometimes provide.

When ERC4337 moves to its P2P mempool, these calculations definitely won't apply, and we will need better tracking of mempool priority fees to give users better hints to get their transaction mined at the lowest cost.